cuML/test/rproj_test.cu

```diff
@@ -0,0 +1,170 @@
+            false,          // gaussian or sparse method
+            -1.0,           // auto density (-1: auto-deduction)
+            false,          // not used
+            42              // random seed
+        };
+
+        std::vector<T> v(N*M);
+        std::generate(v.begin(), v.end(), std::rand);
+        allocate(d_input2, v.size());
+        updateDevice(d_input2, v.data(), v.size(), stream);
+        random_matrix2 = new rand_mat<T>();
+        RPROJfit(random_matrix2, params2);
+        allocate(d_output2, N*params2->n_components);
+        RPROJtransform(d_input2, random_matrix2, d_output2, params2);
+
+        CUDA_CHECK(cudaStreamDestroy(stream));
+    }
+
+ 	void SetUp() override {
+        gaussianTest();
+        sparseTest();
+	}
+
+	void TearDown() override {
+        CUDA_CHECK(cudaFree(d_input1));
+        CUDA_CHECK(cudaFree(d_output1));
+        CUDA_CHECK(cudaFree(d_input2));
+        CUDA_CHECK(cudaFree(d_output2));
+        delete params1;
+        delete random_matrix1;
+        delete params2;
+        delete random_matrix2;
+	}
+
+protected:
+    paramsRPROJ* params1;
+    T* d_input1;
+    rand_mat<T> *random_matrix1;
+    T* d_output1;
+
+    paramsRPROJ* params2;
+    T* d_input2;
+    rand_mat<T> *random_matrix2;
+    T* d_output2;
+};
+
+
+typedef RPROJTest<float> RPROJTestF;
+TEST_F(RPROJTestF, Result) {
+    int M = params1->n_features;
+    int D = params1->n_components;
+    ASSERT_TRUE(D == 10);
+    ASSERT_TRUE(random_matrix1->dense_data);
+
+    float* h_dense_data = new float[M*D];
+    updateHost(h_dense_data, random_matrix1->dense_data, M*D, NULL);
+    ASSERT_TRUE(!std::any_of(h_dense_data, &h_dense_data[M*D-1], [](int i){return i < -1.0 || i > 1.0;}));
+    delete[] h_dense_data;
+
+    int N = params2->n_samples;
+    M = params2->n_features;
+    D = params2->n_components;
+    ASSERT_TRUE(D == 690);
+    ASSERT_TRUE(params2->density == 1/sqrt(M));
+    ASSERT_TRUE(random_matrix2->indices);
+    ASSERT_TRUE(random_matrix2->indptr);
+    ASSERT_TRUE(random_matrix2->sparse_data);
+    ASSERT_TRUE(random_matrix2->sparse_data_size = N*D);
+}
+
+typedef RPROJTest<double> RPROJTestD;
+TEST_F(RPROJTestD, Result) {
+    int M = params1->n_features;
+    int D = params1->n_components;
+    ASSERT_TRUE(D == 10);
+    ASSERT_TRUE(random_matrix1->dense_data);
+
+    double* h_dense_data = new double[M*D];
+    updateHost(h_dense_data, random_matrix1->dense_data, M*D, NULL);
+    ASSERT_TRUE(!std::any_of(h_dense_data, &h_dense_data[M*D-1], [](int i){return i < -1.0 || i > 1.0;}));
+    delete[] h_dense_data;
+
+    int N = params2->n_samples;
+    M = params2->n_features;
+    D = params2->n_components;
+    ASSERT_TRUE(D == 690);
+    ASSERT_TRUE(params2->density == 1/sqrt(M));
+    ASSERT_TRUE(random_matrix2->indices);
+    ASSERT_TRUE(random_matrix2->indptr);
+    ASSERT_TRUE(random_matrix2->sparse_data);
+    ASSERT_TRUE(random_matrix2->sparse_data_size = N*D);
+}
+
```

Comment:
Let's include a simple test here to verify that the distances between all points are actually < epsilon. Perhaps include a couple, for varying n_features and n_components sizes.